File: q2/src/knapsack01.jl
function OptimalSubsolution(W::Int64, w::Vector{Int64}, v::Vector{Int64})
    Rows = fill(0, (2, W+1))

    for i = 1:length(v)
        for c = 0:W
            if c == 0
                Rows[2,c+1] = 0
            elseif w[i] <= c
                Rows[2,c+1] = max(Rows[1,(c+1) - w[i]] + v[i], Rows[1,c+1])
            else
                Rows[2,c+1] = Rows[1,c+1]
            end
        end
        Rows[1,:] = copy(Rows[2,:])
    end

    return Rows[2,:]
end

function Knapsack(W::Int64, i::Vector{Int64}, w::Vector{Int64}, v::Vector{Int64})
    N = length(i)
    if N == 1
        if w[1] <= W
            return [i[1]]
        else
            return []
        end
    end

	il = i[1:div(end,2)]
	ir = i[div(end,2)+1:end]
	wl = w[1:div(end,2)]
	wr = w[div(end,2)+1:end]
	vl = v[1:div(end,2)]
	vr = v[div(end,2)+1:end]

	x1 = OptimalSubsolution(W, wl, vl) # O(nW)
	x2 = OptimalSubsolution(W, wr, vr) # O(nW)

    k = argmax(map(+, x1, reverse(x2))) - 1

	left  = Knapsack(k, il, wl, vl)
    right = Knapsack(W-k, ir, wr, vr)

    return [left ; right]
end

lines = open(readlines, ARGS[1])
W = parse(Int64, match(r"\d+", lines[1]).match)
N = parse(Int64, match(r"\d+", lines[2]).match)

w = fill(0, N)
v = fill(0, N)

for (i,line) in enumerate(lines[3:end])
    matches = collect(eachmatch(r"\d+", line))
    w[i] = parse(Int64, matches[1].match)
    v[i] = parse(Int64, matches[2].match)
 end

@time items = Knapsack(W, collect(1:N),  w, v)
weight = sum(map(i -> w[i], items))
value = sum(map(i -> v[i], items))

function writeresults(f)
	write(f, "V $(value)\n")
	write(f, "i $(length(items)) \n")
	for item in items
		write(f, "$(item)\n")
	end
end

open(writeresults, ARGS[2], "w")

File: q2/src/knapsack01.py
import sys
import re
import numpy as np

filename = sys.argv[1]

f = open(filename, "r")
lines = f.read().split("\n")[:-1]
f.close()

W = int(re.search(r"\d+", lines[0]).group(0))

wtsvals = [list(map(int,re.findall(r"\d+", line))) for line in lines[2:]]
wt = [x[0] for x in wtsvals]
vl = [x[1] for x in wtsvals]

def spaceEfficientKnapsack(W,w,v):
    n = len(v)
    prev = np.arange(W+1)
    prev.fill(0)
    curr = np.arange(W+1)
    curr.fill(0)

    for i in range(1,n+1):
        for j in range(W+1):
            if j == 0:
                curr[j] = 0
            elif w[i-1] <= j:
                curr[j] = max(prev[j-w[i-1]] + v[i-1], prev[j])
            else:
                curr[j] = prev[j]
        prev = [x for x in curr]
    return curr

def optimal_subsolution(W,w,v):
    n = len(w)
    x1 = spaceEfficientKnapsack(W,w[:n//2], v[:n//2])
    x2 = spaceEfficientKnapsack(W,w[n//2:], v[n//2:])
    
    maxval = -1
    maxcap = -1
    for i in range(W+1):
        sm = x1[i] + x2[W-i]
        if sm > maxval:
            maxval = sm
            maxcap = i

    return maxcap

def knapsack_helper(W,i,w,v):
    n = len(i)
    if len(i) == 1:
        if w[0] <= W:
            return [i[0]]
        else:
            return []
    else:
        k = optimal_subsolution(W,w,v)

        il = i[:n//2]
        ir = i[n//2:]
        wl = w[:n//2]
        wr = w[n//2:]
        vl = v[:n//2]
        vr = v[n//2:]
        left = knapsack_helper(k, il, wl, vl)
        right = knapsack_helper(W-k, ir, wr, vr)
        
        return left + right

def knapsack(W,w,v):
    i = list(range(1,len(v)+1))
    return knapsack_helper(W,i,w,v)

f = open("out/sltn.txt", "w")
items = knapsack(W,wt,vl)
val = sum(map(lambda x: vl[x-1], items))
f.write(f"V {val}\n")
f.write(f"i {len(items)}\n")
for i in items:
    f.write(str(i))
    f.write("\n")
f.close()

